platform/rct/hal/usb_stm32/USB-CDC_device_stm32.c

```diff
@@ -1,52 +1,250 @@
+	usb_state._wlock = xSemaphoreCreateMutex();
+
+	if (usb_state._rlock == NULL || usb_state._wlock == NULL) {
+		return -1;
+	}
+
+	Set_System();
+	Set_USBClock();
+
+	USB_Interrupts_Config();
+
+	USB_Init();
 
-static volatile int _init_flag = 0;
-USB_OTG_CORE_HANDLE USB_OTG_dev __attribute__ ((aligned (4)));
+	usb_state._init_flag = 1;
 
-int USB_CDC_device_init()
+	return 0;
+}
+
+void USB_CDC_Write(portCHAR *cByte, int len)
 {
-    vcp_setup();
+	xSemaphoreTake(usb_state._wlock, portMAX_DELAY);
+	while (len--) {
+		/* If the transmit buffer is full, wait until the USB
+		 * hardware has had time to flush it out */
+		uint16_t usb_buf_elements = USB_BUF_ELTS(usb_state.USB_Tx_ptr_in,
+							 usb_state.USB_Tx_ptr_out,
+							 VIRTUAL_COM_PORT_DATA_SIZE);
+
+		while (usb_buf_elements == VIRTUAL_COM_PORT_DATA_SIZE - 1) {
+			portYIELD();
+			usb_buf_elements = USB_BUF_ELTS(usb_state.USB_Tx_ptr_in,
+							usb_state.USB_Tx_ptr_out,
+							VIRTUAL_COM_PORT_DATA_SIZE);
+		}
 
-    /* Initialize the USB hardware */
-    USBD_Init(&USB_OTG_dev,
-              USB_OTG_FS_CORE_ID,
-              &USR_desc,
-              &USBD_CDC_cb,
-              &USR_cb);
-    _init_flag = 1;
+		usb_state.USB_Tx_Buffer[usb_state.USB_Tx_ptr_in] = *cByte++;
 
-    return 0;
+		/* Handle wrapping */
+		if(usb_state.USB_Tx_ptr_in == VIRTUAL_COM_PORT_DATA_SIZE - 1) {
+			usb_state.USB_Tx_ptr_in = 0;
+		} else {
+			usb_state.USB_Tx_ptr_in++;
+		}
+
+	}
+	xSemaphoreGive(usb_state._wlock);
 }
 
+/* TODO: Deprecate this.  It's unused, and having to use strnlen this
+ * deep in the driver stack gives me a wiggins */
 void USB_CDC_send_debug(portCHAR *string)
 {
-    int len = strnlen(string, 2048);
-    vcp_tx((uint8_t*)string, len);
+	if (string == NULL) {
+		return;
+	}
+
+	int len = strnlen(string, 2048);
+
+	USB_CDC_Write(string, len);
 }
 
 void USB_CDC_SendByte( portCHAR cByte )
 {
-    vcp_tx((uint8_t*)&cByte, 1);
+	USB_CDC_Write(&cByte, 1);
 }
 
 portBASE_TYPE USB_CDC_ReceiveByte(portCHAR *data)
 {
-    return vcp_rx((uint8_t*)data, 1, 0);
+	return USB_CDC_ReceiveByteDelay(data, 0);
 }
 
 portBASE_TYPE USB_CDC_ReceiveByteDelay(portCHAR *data, portTickType delay )
 {
-    return vcp_rx((uint8_t*)data, 1, delay);
+	xSemaphoreTake(usb_state._rlock, portMAX_DELAY);
+	int ret = 1;
+	bool byte_received = false;
+
+	while (!byte_received) {
+		if (usb_state.USB_Rx_ptr_out != usb_state.USB_Rx_ptr_in) {
+			*data = usb_state.USB_Rx_Buffer[usb_state.USB_Rx_ptr_out++];
+			/* If we've cleared the buffer, send a signal to the
+			 * USB hardware to stop NAKing packets and to
+			 * continue receiving */
+			if (usb_state.USB_Rx_ptr_out == usb_state.USB_Rx_ptr_in) {
+				/* Enable the receive of data on EP3 */
+				SetEPRxValid(ENDP3);
+			}
+
+			/* Note: We don't need to handle wrapping in
+			 * the receive calls because we always fill
+			 * the rx buffer from the start and drain to completion */
+			byte_received = true;
+		} else {
+			if (delay == portMAX_DELAY) {
```

Comment:
Its just always the largest value you can pass.  In the queues it is designed to mean that, but in practice you should always program as if the lock time expired and be able to handle that corner case.  Altering the behavior here is probably a bad idea so I would rather you be consistent with the rest of the system and handle that super rare case where the timeout expires.  Should be trivial to do...